Refactor/simplify container

ciaranmcnulty opened the pull request:
Bearing in mind there's likely to be some changes in how extensions operate, I've taken the opportunity to simplify the container usage, chiefly to introduce an interface but also to mark areas likely to be volatile as `@internal`.

Having the interface should enable us to change the internally-used container, as well as support more container types in extensions in future via adapters.

The only part of the interface I particularly dislike is the `getByPrefix` which always felt hacky, but it's behaviour should be simple enough to emulate in other containers.

Sam-Burns commented:
Scrutinizer's a bit miffed about coding standards; otherwise, definitely a :+1:

everzet commented:
1. Now as your set/get doesn't transform values when fetching them, you don't need separate setParameter/getParameter
2. define() method name is a bit confusing. I'd prefer factory(), factoryFor(), initialiseVia() or something like that

ciaranmcnulty commented:
@everzet I'm checking that services are objects at the moment... I would find mixing parameters and services a bit weird...

Factory might be better than Definition but there isn't a handy verb for it.

stof commented on src/PhpSpec/Console/Application.php:
```diff
@@ -20,17 +20,19 @@
 /**
  * The command line application entry point
+ *
+ * @Internal
```
should be ``@internal`` in lowercase to match phpdoc

stof replied:
same everywhere

stof commented on src/PhpSpec/Console/Application.php:
```diff
@@ -39,12 +41,12 @@ class Application extends BaseApplication
 
     /**
-     * @return ServiceContainer
+     * @return IndexedServiceContainer
```
shouldn't it keep document the interface ?

ciaranmcnulty replied:
Yep, good catch - I'm considering removing this BTW, it's currently only used for testing? It should be `@internal` anyhow

stof commented on src/PhpSpec/ServiceContainer.php:
```diff
@@ -1,128 +1,51 @@
+     * @param string $id
+     * @param object $value
      *
      * @throws \InvalidArgumentException if service is not an object or callable
```
it should not allow callables anymore, as definitions are handled by a separate method

ciaranmcnulty commented:
I've noticed a performance issue now due to too much `call_user_func` so I'll simplify the container implementation later.

stof commented on src/PhpSpec/ServiceContainer/IndexedServiceContainer.php:
```diff
@@ -0,0 +1,238 @@
+    {
+        foreach ($tags as $tag) {
+            if (!isset($this->tags[$tag])) {
+                $this->tags[$tag] = [];
```
this is useless, due to the way PHP handle nested writes in arrays

ciaranmcnulty replied:
It is? Hm I thought it triggered a warning. that's good then!

stof replied:
reading triggers a warning. but ``$this->tags[$tag][] = $id;`` does not need reading the current value

ciaranmcnulty replied:
Cool, today I learned :-)

stof commented on src/PhpSpec/ServiceContainer/IndexedServiceContainer.php:
```diff
@@ -70,61 +75,42 @@ public function getParam($id, $default = null)
-                    $instance = call_user_func($callable, $container);
-                }
+        $this->definitions[$id] = $value;
+        unset($this->services[$id]);
```
should we allow overwriting a service definition once we have instantiated it ? this can only cause issues (other services may be referencing the old instance as a dependency).
For reference, Pimple 3 forbids this

ciaranmcnulty replied:
That's a very good point. Current services *are* redefined in the behat application suite (which uses Symfony application tester). 

Specifically this happens in scenarios that do describe followed by run.

ciaranmcnulty replied:
BTW this is allowed in the current version, but there's no test covering it.

stof replied:
but are they redefined after instantiation ? If yes, this is fragile (the object graph could use the non-overwritten service). If no, this is not an issue

stof replied:
if we do describe+run, we should probably isolate these 2 calls by recreating the application, because they would not run in the same PHP process in real usage and so they cannot share state

ciaranmcnulty replied:
Yep could be, there'd be some work to refactor the behat suite, which I'll do if there's a clear benefit? Maybe there just needs a `reset` method on the container?

stof commented on src/PhpSpec/ServiceContainer/IndexedServiceContainer.php:
```diff
@@ -70,61 +75,42 @@ public function getParam($id, $default = null)
-    public function define($id, $callable, $tags = [])
+    public function define($id, $value, $tags = [])
     {
-        if (!is_callable($callable)) {
```
this check should not be removed IMO. Definitions must be callables

stof commented on src/PhpSpec/ServiceContainer/IndexedServiceContainer.php:
```diff
@@ -0,0 +1,238 @@
+     *
+     * @throws \InvalidArgumentException if configurator is not a callable
+     */
+    public function addConfigurator($configurator)
```
As you started typehinting the definition callable, you could use a typehint here too

stof commented:
@ciaranmcnulty please look at https://github.com/phpspec/phpspec/pull/961#discussion_r67895606 even if Github collapses it by default due to other changes.
It makes sense to add such restriction to avoid hard-to-debug issues due to a misuse

stof commented:
@ciaranmcnulty your PR changes the behavior of the ``set`` method. I don't know whether it is intended. The old method allows to do 2 things:

- set a service instance directly, which will be reused all the time until replaced. In the Symfony DIC, this is a shared (container-scoped in 2.x) synthetic service
- set a service definition which will rebuild the service each time. In Symfony DIC, this is a non-shared (prototyped-scoped in 2.x) service

``setShared`` was used to created shared services, which is what ``define`` does know, and is fine.

You changed most places using ``set`` to use ``define``, but this changes the behavior. Most changes are fine (it makes sense to keep it shared in most cases, as using set for many of them was just a performance optimization relying on the fact that they were used by a single shared service, but optimizing the way services are shared instead is way better), but changes should be reviewed carefully.
And we need to decide whether all services should be shared or no (knowing whether we need a replacement feature or no)

ciaranmcnulty commented:
Yes indeed, there were three methods:

* `set` with object - object storage - now `set`
* `setShared` with callable - factory - now `define`
* `set` with callable - prototype-style factory - now removed

Why remove the third? As I went through the internal uses of `set` with a callable, I didn't find any that needed to be prototype-style, so I dropped support for it on the basis that it kept things simple.

I don't think non-shared services are, in general, a good idea* but I could add a third method (definePrototype?) that supports it if there's a good case for extensions using it? I definitely don't want to keep the overloaded `set` because of the ambiguity when you `set` an object that has an `__invoke`.

* Actually if they're stateless it's fine, but that's not the style PhpSpec is currently using.

stof commented:
@ciaranmcnulty I'm totally fine with removing support for the prototype scope, as long as we don't need it in Phpspec. We just need to make sure we don't need it

ciaranmcnulty commented:
I'm 95% confident all the places it changed in this PR, `describe` was appropriate. I haven't audited extensions but presumably when they add `3.0` support they'll come and tell us and we'll need some BC-compatible way to add stuff.
